### DISCUSSION TITLE
IPtables Rules

### DISCUSSION THREAD

**Discussion by andrew-hoff** [strongswan/strongswan#829] [Q&A] [answered]
I have a setup where I use transport mode, GRE tunnels and a route based VPN. All of this is configured using the vici interface. In this scenario, if the IPSec policy is torn down, my GRE tunnel may still be up along with the corresponding route. As a result, packets may go through unencrypted. To mitigate this, I added the following two iptables rules:

- `-t filter -A OUTPUT -m policy --dir out --pol none -p gre -j DROP`
- `-t filter -A OUTPUT -m policy --dir in --pol none -p gre -j DROP`

The objective here is to ensure that all GRE packets are subject to IPSec processing (I still have non-GRE interfaces that I want to be able to use without IPSec). Any GRE packets that are not subject to IPSec processing are dropped. This way, if there is a GRE tunnel with no corresponding policy + SA, packets can't be sent through it without being dropped.

My concern is if a race condition is possible here. Let's say a packet is going through netfilter and does not match these rules because it is subject to IPSec processing i.e. a policy exists. However, immediately after that, the IPSec policy and SA are dropped. At this point, the packet goes through XFRM and is not encrypted. Is this possible? If so, is there a better way for me to ensure that only GRE packets that are subject to encryption can leave the device? Should I be doing this at the policy level instead?

Another idea I had was policies like this:
- `ip xfrm policy add proto gre dir in action block priority 100`
- `ip xfrm policy add proto gre dir out action block priority 100`

**Comment by andrew-hoff** [comment]
This [link](https://wiki.strongswan.org/projects/strongswan/wiki/SecurityRecommendations#Tunnel-Shunting) implies that either approach can work (though I'll need to increase my priority value). Is either preferred?

**Reply by Thermi** [reply]
You can just configure a drop policy in strongSwan with a corresponding priority to take care of it. There's no need to execute any `ip xfrm` commands.

**Comment by andrew-hoff** [comment, marked as answer]
Is that done by setting the mode to `drop` [here](https://manpages.debian.org/testing/strongswan-swanctl/swanctl.conf.5.en.html#connections._conn_.children._child_.mode) along with corresponding priority [here](https://manpages.debian.org/testing/strongswan-swanctl/swanctl.conf.5.en.html#connections._conn_.children._child_.priority)?

**Reply by Thermi** [reply]
Yes.
